### DISCUSSION TITLE
Starting to get discouraged.

### DISCUSSION THREAD

**Discussion by tausyy** [AvaloniaUI/Avalonia#6165] [General]
I'm trying to move away from WinForms which I've been using for hobby project since 2002. Trying to use MVVM and Avalonia/ReactiveUI to very hard for me because I can't find any good resources for more intermediate help. I've already worked through the two tutorials TodoList and MusicStoreApp. Where else should I go? MVVM really seems like an overly complicated way to write apps. I just want to open a dialog that returns a true or false value and every example I look at is beyond my comprehension. Halp, pls. Point me somewhere.

**Comment by wieslawsoltes** [comment]
This is some basic example of how you can do dialog and other custom navigation elements.

https://github.com/wieslawsoltes/NavigationSample

**Comment by wieslawsoltes** [comment]
Another demo project with different solution, check how the Settings button displays dialog

https://github.com/wieslawsoltes/WalletWasabiSandbox

**Reply by tausyy** [reply]
so what it looks like to me with this one. it doesn't actually open up another window, rather, it displays a hidden "dialog" panel?

**Comment by wieslawsoltes** [comment]
Another solution for dialogs

https://github.com/wieslawsoltes/Core2D

https://github.com/wieslawsoltes/Core2D/blob/e467fb1ae3ee9010d62de9bb0df6dd4f69ca0615/src/Core2D/Views/MainView.axaml#L28-L39

https://github.com/wieslawsoltes/Core2D/blob/master/src/Core2D/Model/Editor/IDialogPresenter.cs

https://github.com/wieslawsoltes/Core2D/blob/master/src/Core2D/ViewModels/Editor/DialogViewModel.cs

https://github.com/wieslawsoltes/Core2D/blob/master/src/Core2D/Model/Editor/IDialogPresenter.cs

https://github.com/wieslawsoltes/Core2D/blob/master/src/Core2D/Views/DialogView.axaml

**Comment by wieslawsoltes** [comment]
This one is bit more complex as it's part of larger application:

https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Fluent/Views/MainView.axaml#L84-L97

https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Fluent/Controls/Dialog.axaml

https://github.com/zkSNACKs/WalletWasabi/tree/master/WalletWasabi.Fluent/ViewModels/Dialogs/Base

**Comment by tausyy** [comment]
I appreciate the examples and I will look at them but that only shows me what is happening but doesn't explain why or what. I cant find any resources explaining these concepts.

**Comment by maxkatz6** [comment]
> MVVM really seems like an overly complicated way to write apps

You are not required to write in MVVM, if it's your app. And in general I would recommend looking on "getting started" tutorials for WPF and UWP as well, since they are very similar with Avalonia in many ways.

> ReactiveUI 

ReactiveUI is by itself very complicated, but still very powerful tool. Some people might prefer [Prism](https://github.com/AvaloniaCommunity/Prism.Avalonia) instead. In both cases for documentation and tutorials you can look into WPF related topics, these frameworks reuse same concepts ported to different platforms.

> I just want to open a dialog that returns a true or false value

Simplest way what I see is to create Window with boolean property. Show this window with Show() or ShowDialog() method. And inside of this window handle button click event to set that boolean property to true or false. After window was closed read this property. It's not MVVM way, but it's simple and similar to way how it could be done in WinForms.

For more complex scenarios you should take a look on examples provided by @wieslawsoltes . Also you can take a look on this page https://github.com/AvaloniaUI/Avalonia/discussions/5344. Basically you can define interface to open dialog, which will be used in the ViewModel layer, and implemented in View layer. Why? In that way your View Model will not know details of implementation from View layer (you can even avoid any Avalonia/WPF/UWP specific references from it, making this layer portable). Also it make it easy to mock this service to unit test or make e2e tests. Although ReactiveUI Interactions will do the same, but a bit simpler.
